### PR TITLE
Refactor around FEMContext::some_foo

### PR DESCRIPTION
**Pull request by roystgnr**
This saves a little code, but more importantly it adds implementations
of a few obscure methods like side_rate().

This passes the GRINS test suite, which ought to exercise this code fairly well.

### CODE REVIEW

**Comment by pbauman** [comment]
Dat pointer-to-member function. :+1:

**Comment by roystgnr** [comment]
Pointer-to-member function overloaded by a templated argument.  It took me three tries and one round of "let's see if clang gives more comprehensible compiler warnings" to get that to compile.

I'll merge this today unless anyone screams.  I'm fairly confident our tests would have caught any bugs, though.  Aside from FEMSystem and GRINS stuff, the refactored System projection code uses FEMContext now, so there's code coverage from all directions.

**Comment by permcody** [comment]
That has got to be the craziest non-obfuscated C++ code I have ever looked
at! Pretty cool when it's all said and done!

**Comment by jwpeterson** [comment]
> non-obfuscated C++ code
```
<OutputType>::type>::type>::type>
```

File: src/systems/fem_context.C [hunk 1]
```diff
@@ -250,8 +259,16 @@ void FEMContext::some_interior_gradient(unsigned int var, unsigned int qp, Outpu
 
 #ifdef LIBMESH_ENABLE_SECOND_DERIVATIVES
 template<typename OutputType,
+         void (FEMContext::*fe_getter)
+           (unsigned int,
+            FEGenericBase
+              <typename TensorTools::MakeReal
+                <typename TensorTools::DecrementRank
+                  <typename TensorTools::DecrementRank
+                    <OutputType>::type>::type>::type> *&)
```
Comment (jwpeterson) [inline src/systems/fem_context.C]: Is it possible to use a `typedef` for the function pointer type?  Seems like you have repeated the same 7 lines of code used from line 843 of the header file...

Comment (roystgnr) [inline src/systems/fem_context.C]: It's possible with a C++11 using statement, but not in C++03.  OutputType is a free argument and appears as a template argument to DecrementRank here, and "template typedef" doesn't exist.  Best we could do is to add a nested templated class to the header and use three typedefs inside that class.  I'll bet that ends up being just as many lines of code, but I'll try it out if you like.

Comment (jwpeterson) [inline src/systems/fem_context.C]: Hmm... maybe a templated helper class (basically a c++03-compatible template typedef) would be good here?

I (tried)[https://github.com/jwpeterson/libmesh/commit/42fc9791d287f697c5c10e217968c07e01b67a40] to add one, but it doesn't compile.  Not sure if it's something fundamental or just a syntax error.

Comment (jwpeterson) [inline src/systems/fem_context.C]: Pushed one more commit to the [branch](https://github.com/jwpeterson/libmesh/tree/fem_context_refactor) which makes it compile.  The net lines are the same, but I think this follows the DRY principle a bit better.

**Comment by roystgnr** [comment]
What do you think of this version?  Sorry, didn't mean to step on your toes; I'd already half-finished it when I saw you were working on the same.

**Comment by jwpeterson** [comment]
:+1:

**Comment by roystgnr** [comment]
Cool.  I'll merge once the MooseBuild and GRINS tests all finish.

**Pull request merged**